src/widget.ts

```diff
@@ -101,4 +127,49 @@ export class VegaWidget extends DOMWidgetView {
     // initial rendering
     await reembed();
   }
+
+  updateDataFrame(): any[] {
+    let table = this.model.get("_df");
+    // console.log("table", table);
+    const proxy = rowProxy(table);
+    const rows = Array(table.size);
+
+    for (let i = 0, n = rows.length; i < n; ++i) {
+      rows[i] = proxy(i);
+    }
+    return rows;
+    // for(let i=0; i < table.size; i++){
+    //     let row: any = [];
+    //     for (const col of table.columns){
+    //         if(table.data[col].shape===undefined){
+    //             row[col] = table.data[col][i];
+    //         } else {
+    //             row[col] = table.data[col].get(i);
+    //         }
+    //     }
+    //     res[i] = row;
+    // }
+    // return res;
+  }
+
+  updateArray2D(): any[] {
+    // console.log("updateArray2D");
+    let table = this.model.get("_df");
+    let res = Array(table.size * table.size);
+    let fancyCol = table.columns[0];
+    let arr: ndarray.NdArray = table.data[fancyCol];
+    let cols: string[] = fancyCol.split(",");
+    //let cols = this.model.get("_columns");
+    let k = 0;
+    for (let i = 0; i < arr.shape[0]; i++) {
```

Comment:
What is this code doing? I suspect this can be done much more efficiently with some methods in ndarray.

Reply:
I'm still unclear about this.